### ISSUE TITLE
`Error: out of range source coordinates for image copy` with for map icon sprite

### ISSUE THREAD

**Issue by tordans** [maplibre/maplibre-gl-js#575]
**maplibre-gl-js version**:

`"maplibre-gl": "^1.15.2",`

**browser**:

Firefox

### Steps to Trigger Behavior

Open map in Firefox. ~ 50 % of the time, the map icons do not load. Instead, I see all labels on top of eachother.
In Chrome, it happens less often (a lot less).
The second reload is usually fine. But following shift-reloads can re-trigger the error (but not often).

The easiest way to reproduce it, is to constantly close-and-open private browsing windows.

### Link to Demonstration

I created a temporary demo page at https://tordans.github.io/flyingsparks-blog/maplibre-sprite-debugging-map.html
([Github](https://github.com/tordans/flyingsparks-blog/blob/main/maplibre-sprite-debugging-map.html))

### Expected Behavior

The map icons should show all the time. And if possible, the error message should tell me what to do …

### Actual Behavior

| Expected | Error |
|---|---|
| <img width="303" alt="good" src="https://user-images.githubusercontent.com/111561/138944797-4151841d-2dc2-440d-b1b9-56453a8f135b.png"> | <img width="385" alt="error" src="https://user-images.githubusercontent.com/111561/138944801-33f6a0d7-9e41-4b3f-ab45-8511ad75ff76.png"> |

Error message:

<img width="609" alt="error-message" src="https://user-images.githubusercontent.com/111561/138944784-4e2cc16a-847d-4c3e-b21a-a3e7d1a4db53.png">

<details>
<summary><code>Error: out of range source coordinates for image copy</code></summary>

```
Error: out of range source coordinates for image copy
    p ajax.js:183
    promise callback*p ajax.js:182
    l ajax.js:151
    promise callback*l ajax.js:148
    vt ajax.js:190
    bt ajax.js:257
    getJSON ajax.js:268
    _spriteRequest load_sprite.js:19
    _loadSprite style.js:285
    _load style.js:257
    _request style.js:225
    p ajax.js:181
    promise callback*p ajax.js:170
    l ajax.js:151
    promise callback*l ajax.js:148
    vt ajax.js:190
    bt ajax.js:257
    getJSON ajax.js:268
    loadURL style.js:220
    _updateStyle map.js:1386
    setStyle map.js:1341
    r map.js:466
    <anonymous> maplibre-sprite-debugging-map.html:17
evented.js:140:20
```

</details>

### Other information

- We are using MapTiler with [Maptiler Desktop](https://www.maptiler.com/desktop/) to create the map icon layer.
- The sprites are manually created within maptiler.
  - Sprite json: `{"station-mid@1x":{"width":45,"height":50,"x":0,"y":0,"pixelRatio":1},"station-mid@2x":{"width":45,"height":50,"x":45,"y":0,"pixelRatio":1}}`
  - Sprite file: ![image](https://user-images.githubusercontent.com/111561/138945952-d20b2a61-de8f-4eea-aef4-975d22224c1d.png)
- The sprite files (json, png) are always visible without any error message in the network tab, even in the broken view (so it's not that they don't load…)
- The only related issue I could find is https://github.com/mapbox/mapbox-gl-js/issues/5860#issuecomment-351793206, which is why I re-created the sprites (json and png) manually; that did not help. (Note, we might change the sprites again in the next few days.)

**Comment by clementgayvallet** [comment]
I investigated the issue for firefox and after some logging I found this error happens when the sprite index response is received **after** the sprite image.

There are 2 reasons for this:
1. When the sprite image is received first, in the [load_sprite](https://github.com/maplibre/maplibre-gl-js/blob/main/src/style/load_sprite.ts#L11) function, the image reference is kept and the maybeComplete() function just do nothing. Then when the sprite index is received the maybeComplete() function loads the sub-images.
2. However, as you can see in the [arrayBufferToImage](https://github.com/maplibre/maplibre-gl-js/blob/9d1381ce0492c6703934abb51fb4fbee8310d038/src/util/ajax.ts#L299) function (called by the [getImage](https://github.com/maplibre/maplibre-gl-js/blob/9d1381ce0492c6703934abb51fb4fbee8310d038/src/util/ajax.ts#L333) function), on browsers where an HTMLImageElement is used to load the image (currently FF and Safari), a transparent pixel hack is used to prevent memory leaks.

Here are some details about what is going on:
1. getJSON and getImage are called
2. getImage responds first and loads the data into an HTMLImageElement
3. The [callback](https://github.com/maplibre/maplibre-gl-js/blob/9d1381ce0492c6703934abb51fb4fbee8310d038/src/util/ajax.ts#L302) resolves which keeps the image reference and calls maybeComplete() which does nothing
4. Once the callback returns the [transparent pixel hack](https://github.com/maplibre/maplibre-gl-js/blob/9d1381ce0492c6703934abb51fb4fbee8310d038/src/util/ajax.ts#L333) runs, so the image reference above now references a transparent pixel
5. Later the getJSON responds and calls maybeComplete() which loads sub-images... but from a transparent pixel, hence the error.

Most of the time the sprite image and the sprite index are in the browser cache so the requests respond almost instantly and in order i.e. the index then the image so the error doesn't happen. But when the console is open and the cache disabled, the error may happen.

This PR #650 should fix this problem on FF by using an ImageBitmap instead of an HTMLImageElement (and its transparent pixel hack). However the problem will remain on Safari.

**Comment by HarelM** [comment]
I've seen the maybe completed code. It doesn't look very good in my opinion.
Maybe changing it to use promises might be a better approach instead of using a transparent pixel...?

**Comment by xabbu42** [comment]
Given that #650 was merged, should this be closed?

**Comment by tordans** [comment]
>Given that https://github.com/maplibre/maplibre-gl-js/pull/650 was merged, should this be closed?

Yes, let's close it and maybe reopen should the issue re-surface.
